hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestScanMetricsByRegion.java

```diff
@@ -0,0 +1,170 @@
+import org.apache.hadoop.hbase.client.metrics.ScanMetrics;
+import org.apache.hadoop.hbase.testclassification.ClientTests;
+import org.apache.hadoop.hbase.testclassification.MediumTests;
+import org.apache.hadoop.hbase.util.Pair;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+@Category({ ClientTests.class, MediumTests.class })
+public class TestScanMetricsByRegion extends FromClientSideBase {
+  @ClassRule
+  public static final HBaseClassTestRule CLASS_RULE =
+    HBaseClassTestRule.forClass(TestScanMetricsByRegion.class);
+
+  @Rule
+  public TableNameTestRule name = new TableNameTestRule();
+
+  private static final HBaseTestingUtil TEST_UTIL = new HBaseTestingUtil();
+
+  @Parameters(name = "{index}: scanner={0}")
+  public static List<Object[]> params() {
+    return Arrays.asList(new Object[] { "ForwardScanner", new Scan() },
+      new Object[] { "ReverseScanner", new Scan().setReversed(true) });
+  }
+
+  @Parameter(0)
+  public String scannerName;
+
+  @Parameter(1)
+  public Scan originalScan;
+
+  @BeforeClass
+  public static void setUp() throws Exception {
+    // Start the minicluster
+    TEST_UTIL.startMiniCluster(2);
+  }
+
+  @AfterClass
+  public static void tearDown() throws Exception {
+    TEST_UTIL.shutdownMiniCluster();
+  }
+
+  private Scan getScan(byte[][] ROWS, boolean isSingleRegionScan) throws IOException {
+    Scan scan = new Scan(originalScan);
+    if (isSingleRegionScan) {
+      scan.withStartRow(ROWS[0], true);
+      scan.withStopRow(ROWS[0], true);
+    } else if (scan.isReversed()) {
+      scan.withStartRow(ROWS[1], true);
+      scan.withStopRow(ROWS[0], true);
+    } else {
+      scan.withStartRow(ROWS[0], true);
+      scan.withStopRow(ROWS[1], true);
+    }
+    return scan;
+  }
+
+  @Test
+  public void testScanMetricsByRegion() throws Exception {
+    final TableName tableName = name.getTableName();
+    byte[][] ROWS = makeN(ROW, 2);
+    byte[][] FAMILIES = makeNAscii(FAMILY, 2);
+    byte[][] QUALIFIERS = makeN(QUALIFIER, 1);
+    byte[][] VALUES = makeN(VALUE, 2);
+    byte[][] splitKeys = new byte[][] { ROWS[1] };
+    try (Table ht = TEST_UTIL.createTable(tableName, FAMILIES, splitKeys)) {
+      // Add two rows in two separate regions
+      Put put1 = new Put(ROWS[0]);
+      put1.addColumn(FAMILIES[0], QUALIFIERS[0], VALUES[0]);
+      Put put2 = new Put(ROWS[1]);
+      put2.addColumn(FAMILIES[0], QUALIFIERS[0], VALUES[1]);
+      List<Put> puts = Arrays.asList(put1, put2);
+      ht.batch(puts, null);
+      // Verify we have two regions
+      Admin admin = TEST_UTIL.getAdmin();
+      Assert.assertEquals(2, admin.getRegions(tableName).size());
+
+      // Test scan metrics disabled
```

Comment:
nit: each of these little blocks could be its own test method, and then this test would be a lot nicer to reason about when there are failures.